pkg/cmd/readyz/readyz.go

```diff
@@ -0,0 +1,189 @@
+	fs.StringVar(&r.clientKeyFile, "key-file", "", "Health probe TLS client key file. (required)")
+	fs.StringVar(&r.clientCACertFile, "cacert-file", "", "Health probe TLS client CA certificate file. (required)")
+}
+
+// Validate verifies the inputs.
+func (r *readyzOpts) Validate() error {
+	if len(r.targetEndpoint) == 0 {
+		return fmt.Errorf("missing required flag: --target")
+	}
+	if len(r.clientKeyFile) == 0 {
+		return errors.New("missing required flag: --key-file")
+	}
+	if len(r.clientCertFile) == 0 {
+		return errors.New("missing required flag: --cert-file")
+	}
+	if len(r.clientCACertFile) == 0 {
+		return errors.New("missing required flag: --cacert-file")
+	}
+
+	return nil
+}
+
+// Run contains the logic of the readyz command which checks the health of the etcd member
+func (r *readyzOpts) Run() error {
+	shutdownCtx, cancel := context.WithCancel(context.Background())
+	shutdownHandler := server.SetupSignalHandler()
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/readyz", r.getReadyzHandlerFunc(shutdownCtx))
+	// Handle the /healthz endpoint as well since the static pod controller's guard pods check the /healthz endpoint
+	// https://github.com/openshift/library-go/blob/edab248e63516c65a93467eaa8224c86d69f5de9/pkg/operator/staticpod/controller/guard/manifests/guard-pod.yaml#L44
+	mux.HandleFunc("/healthz", r.getReadyzHandlerFunc(shutdownCtx))
+
+	addr := fmt.Sprintf("0.0.0.0:%d", r.listenPort)
+	klog.Infof("Listening on %s", addr)
+
+	server := &http.Server{
+		Addr:        addr,
+		Handler:     mux,
+		BaseContext: func(_ net.Listener) context.Context { return shutdownCtx },
+	}
+	go func() {
+		defer cancel()
+		<-shutdownHandler
+		klog.Infof("Received SIGTERM or SIGINT signal, shutting down server.")
+		server.Shutdown(shutdownCtx)
+	}()
+
+	c := net.ListenConfig{}
+	c.Control = permitAddressReuse
+	ln, err := c.Listen(shutdownCtx, "tcp", addr)
+	if err != nil {
+		return err
+	}
+	err = server.ServeTLS(ln, r.clientCertFile, r.clientKeyFile)
+	if err == http.ErrServerClosed {
+		err = nil
+	}
+	<-shutdownCtx.Done()
+	return err
+}
+
+func (r *readyzOpts) getReadyzHandlerFunc(ctx context.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		etcdClient, err := r.newETCD3Client(ctx, r.targetEndpoint)
+		if err != nil {
+			klog.V(2).Infof("failed to establish etcd client: %v", err)
+			http.Error(w, fmt.Sprintf("failed to establish etcd client: %v", err), http.StatusServiceUnavailable)
+			return
+		}
+
+		// linearized request to verify health of member
+		// TODO: Once learner members are supported, update this to differentiate
+		// learner members and only issue a serialized Get() for them
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		_, err = etcdClient.Get(ctx, "health")
+		if err != nil {
+			klog.V(2).Infof("failed to get member health key: %v", err)
+			http.Error(w, fmt.Sprintf("failed to get member health key: %v", err), http.StatusServiceUnavailable)
+			return
```

Comment:
Explicitly write the failing response back please, along with the detail we got from etcd if we got any.  If we didn't get a body, explain that too.

Reply:
Sorry but could you please expand on what you mean by "explicitly write".
The above `http.Error(w, ...)` is writing the full error response from the etcd client Get() request to the http response writer, which shows up in the response body.
https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/net/http/server.go;l=2093-2098

Just double checked the events for a failing probe and the error does show up in the body, e.g:
```console
$ kubectl -n openshift-etcd describe pod etcd-ip-10-0-223-164.us-west-1.compute.internal
Events:
  Type     Reason      Age   From     Message
  ----     ------      ----  ----     -------
...
  Normal   Created     48s   kubelet  Created container etcd-readyz
  Normal   Started     48s   kubelet  Started container etcd-readyz
  Warning  ProbeError  42s   kubelet  Readiness probe error: HTTP probe failed with statuscode: 503
body: failed to establish etcd client: context deadline exceeded
```

Seems like our kublet doesn't swallow the body.